### PR TITLE
Adds new dotnet new template arg --use-https-redirect

### PR DESCRIPTION
**Pull request by warrenbuckley**
Adds new feature to dotnet new template that was required by the Cloud team, as UmbracoID when cloned down locally from Cloud requires the redirect flow to be running on HTTPS and not HTTP

This new flag adds the following to startup.cs
`app.UseHttpsRedirection();`

## Docs from MS Templating Repo for dotnet new and conditionals in .CS files
https://github.com/dotnet/templating/wiki/Conditional-processing-and-comment-syntax#language-source-files

----

To Test (at least this is how I did it):

- Check out the PR and amend your local `src\Directory.Build.props` file to create a new version, e.g. `<InformationalVersion>9.0.0-rc003abc</InformationalVersion>`
- Run `build\build.ps`
- Copy the resulting `Umbraco.Templates.9.0.0-rc003abc.nupkg` from `build.out` into a local folder set up as a NuGet feed
- Run:
- Verify that the Startup.cs class does NOT include `app.UseHttpsRedirection();` 
- Ensure browsing to site on HTTP does NOT auto redirect
```
dotnet new -i Umbraco.Templates::9.0.0-rc003abc
dotnet new umbraco --name NoRedirect
cd NoRedirect
code .
dotnet run
```

- Retest with the following and verify  `app.UseHttpsRedirection();`  is in Startup.cs
- Verify that the Startup.cs class DOES include `app.UseHttpsRedirection();` 
- Ensure browsing to site on HTTP DOES auto redirect to HTTPS

```
dotnet new umbraco --name WithRedirect --use-https-redirect
cd WithRedirect
code .
dotnet run
```

### CODE REVIEW

File: build/build.ps1 [hunk 1]
```diff
@@ -351,8 +351,8 @@
     $this.CopyFiles("$templates", "*", "$tmp\Templates")
 
     Write-Host "Copy files for dotnet templates"
+    #$this.CopyFiles("$src\Umbraco.Web.UI", "Startup.cs", "$tmp\Templates\UmbracoProject")
```
Comment (p-m-j) [inline build/build.ps1:354]: intentional?

**Comment by ronaldbarendse** [comment]
I don't see a problem with adding the `#if (UseHttpsRedirect)` to the Startup.cs file in the Umbraco.Web.UI project and copying that to the template: that will remove the need to keep both files synced and will just ignore the code within this preprocessor directive (as this symbol is not set).

**Comment by warrenbuckley** [comment]
@ronaldbarendse & @p-m-j all good now. Have you tested it? then I could do with it being merged in so Cloud team can consume it for creating new Umbraco 9 sites on Cloud.

File: src/Umbraco.Web.UI/Startup.cs [hunk 2]
```diff
@@ -54,6 +54,10 @@ public void ConfigureServices(IServiceCollection services)
         /// </summary>
         public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
         {
+#if(UseHttpsRedirect)
```
Comment (ronaldbarendse) [inline src/Umbraco.Web.UI/Startup.cs:57]:
```suggestion
#if (UseHttpsRedirect)
```

File: build/build.ps1 [hunk 3]
```diff
@@ -351,8 +351,8 @@
     $this.CopyFiles("$templates", "*", "$tmp\Templates")
 
     Write-Host "Copy files for dotnet templates"
-    $this.CopyFiles("$src\Umbraco.Web.UI", "Program.cs", "$tmp\Templates\UmbracoProject")
     $this.CopyFiles("$src\Umbraco.Web.UI", "Startup.cs", "$tmp\Templates\UmbracoProject")
+    $this.CopyFiles("$src\Umbraco.Web.UI", "Program.cs", "$tmp\Templates\UmbracoProject")
     $this.CopyFiles("$src\Umbraco.Web.UI\Views", "*", "$tmp\Templates\UmbracoProject\Views")
```
Comment (ronaldbarendse) [inline build/build.ps1:356]: Lets keep the original (alphabetical) order:
```suggestion
    $this.CopyFiles("$src\Umbraco.Web.UI", "Program.cs", "$tmp\Templates\UmbracoProject")
    $this.CopyFiles("$src\Umbraco.Web.UI", "Startup.cs", "$tmp\Templates\UmbracoProject")
    $this.CopyFiles("$src\Umbraco.Web.UI\Views", "*", "$tmp\Templates\UmbracoProject\Views")
```

**Review by ronaldbarendse — CHANGES_REQUESTED** [review]

**Comment by bergmania** [comment]
I guess it should also specific `Umbraco:CMS:Global:UseHttps=true`?

**Comment by nul800sebastiaan** [comment]
@bergmania That would have to be added to the appsettings file(s) when Cloud provisions the site so not part of this PR I'd say.

**Comment by ronaldbarendse** [comment]
@bergmania The `UseHttps` setting is now also added when using the HTTPS redirection 👍🏻

I've tested with the following commands and it produces the correct `appsettings.json` and `Startup.cs`:
```
dotnet new umbraco --name NoRedirect
dotnet new umbraco --name WithRedirect --use-https-redirect
dotnet new umbraco --name WithRedirectNoNodes --use-https-redirect --no-nodes-view-path "~/umbraco/UmbracoWebsite/NoNodes.cshtml"
```

------------------

Setting up a custom local NuGet feed isn't required, as you can also specify the local path during install:
```
dotnet new -i Umbraco.Templates::9.0.0-rc003abc --nuget-source "FullLocalPathTo\build.out"
```

File: build/templates/UmbracoProject/.template.config/ide.host.json [hunk 4]
```diff
@@ -62,6 +62,13 @@
                 "text": "Optional: Path to a custom view presented with the Umbraco installation contains no published content"
             },
             "isVisible": "true"
+        },
+        {
+            "id": "UseHttpsRedirect",
+            "name": {
+                "text": "Optional: Adds code to startup.cs to redirect HTTP to HTTPS"
```
Comment (ronaldbarendse) [inline build/templates/UmbracoProject/.template.config/ide.host.json:69]:
```suggestion
                "text": "Optional: Adds code to Startup.cs to redirect HTTP to HTTPS and enables the UseHttps setting."
```

File: build/templates/UmbracoProject/.template.config/template.json [hunk 5]
```diff
@@ -293,6 +293,12 @@
         "UsingUnattenedInstall":{
             "type": "computed",
             "value": "(FriendlyName != \"\" && Email != \"\" && Password != \"\" && ConnectionString != \"\")"
+        },
+        "UseHttpsRedirect":{
+            "type": "parameter",
+            "datatype":"bool",
+            "defaultValue": "false",
+            "description": "Adds code to startup.cs to redirect HTTP to HTTPS (Default: false)"
```
Comment (ronaldbarendse) [inline build/templates/UmbracoProject/.template.config/template.json:301]:
```suggestion
            "description": "Adds code to Startup.cs to redirect HTTP to HTTPS and enables the UseHttps setting (Default: false)"
```

**Review by ronaldbarendse — APPROVED** [review]

**Pull request merged**